### ISSUE TITLE
Misuses of cryptographic APIs

### ISSUE THREAD

**Issue by mhp0rtal** [egzosn/pay-java-parent#39]
Hi
The following lines have cryptographic API misuses. <br> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/SecureUtil.java: Line number => 156: API name => Signature:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 115: API name => Signature:First parameter (with value "SHA1WithRSA") should be any of {NONEwithDSA, SHA1withDSA, SHA224withDSA, SHA256withDSA, SHA256withRSA, SHA256withECDSA}<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 280: API name => Cipher:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 118: API name => Signature:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/SecureUtil.java: Line number => 44: API name => MessageDigest:Unexpected call to method reset on object of type java.security.MessageDigest. Expect a call to one of the following methods digest,update<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/SignUtils.java: Line number => 68: API name => SecretKeySpec:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 273: API name => Cipher:Operation on object of type javax.crypto.Cipher object not completed. Expected call to update, doFinal<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 139: API name => Signature:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/SecureUtil.java: Line number => 159: API name => Signature:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 46: API name => Signature:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 280: API name => Cipher:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/SecureUtil.java: Line number => 167: API name => Signature:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 181: API name => Cipher:Operation on object of type javax.crypto.Cipher object not completed. Expected call to <javax.crypto.Cipher: int doFinal(byte[],int,int,byte[])>, <javax.crypto.Cipher: byte[] doFinal(byte[])>, <javax.crypto.Cipher: int doFinal(java.nio.ByteBuffer,java.nio.ByteBuffer)>, update, <javax.crypto.Cipher: byte[] doFinal(byte[],int,int)>, <javax.crypto.Cipher: int doFinal(byte[],int,int,byte[],int)><BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 200: API name => Cipher:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 68: API name => Signature:First parameter (with value "SHA1WithRSA") should be any of {NONEwithDSA, SHA1withDSA, SHA224withDSA, SHA256withDSA, SHA256withRSA, SHA256withECDSA}<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/SecureUtil.java: Line number => 164: API name => Signature:First parameter (with value "SHA1withRSA") should be any of {NONEwithDSA, SHA1withDSA, SHA224withDSA, SHA256withDSA, SHA256withRSA, SHA256withECDSA}<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/http/HttpRequestTemplate.java: Line number => 118: API name => KeyStore:Second parameter should never be of type java.lang.String.<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 71: API name => Signature:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 200: API name => Cipher:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 41: API name => Signature:First parameter (with value "SHA1WithRSA") should be any of {NONEwithDSA, SHA1withDSA, SHA224withDSA, SHA256withDSA, SHA256withRSA, SHA256withECDSA}<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/CertDescriptor.java: Line number => 169: API name => KeyStore:Operation on object of type java.security.KeyStore object not completed. Expected call to load<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/SignUtils.java: Line number => 68: API name => SecretKeySpec:<BR> File name => pay-java-common/src/main/java/com/egzosn/pay/common/util/sign/encrypt/RSA.java: Line number => 136: API name => Signature:First parameter (with value "SHA1WithRSA") should be any of {NONEwithDSA, SHA1withDSA, SHA224withDSA, SHA256withDSA, SHA256withRSA, SHA256withECDSA}<BR>

**Comment by egzosn** [comment]
....

**Comment by mhp0rtal** [comment]
Hi,
Thanks for your attention.
I reported the cryptographic misuses due to my recent research on how developers employ Java Cryptography Architecture (JCA) APIs in Java applications. I utilized some static analysis tools intended for checking cryptographic practices. The aforementioned misuses in those files were obtained from the results of the tools. The severity of the bug is relative. For instance, Even though MD5 is listed on Java documentation, MD5 is not recommended to use in any language anymore as it is considered insecure. An attacker can try billions of candidate passwords per second on a single GPU. Nevertheless, depending on the scenario the same hash algorithm may not be a problem for file integrity purpose.
There are several instances of cryptographic misuses which could lead to a security vulnerability:
*Do not use ECB mode for encryption.
*Do not use a non-random IV for CBC encryption
*Do not use constant encryption keys.
*Do not use constant salts for PBE.
*Do not use fewer than 1,000 iterations for PBE.
*Do not use static seeds to seed SecureRandom(·)
